gh-83938, gh-122476: Stop incorrectly RFC 2047 encoding non-ASCII email addresses

medmunds opened the pull request:
This PR fixes gh-83938 and fixes gh-122476, which have the same underlying issue.

Email generators had been incorrectly flattening non-ASCII email addresses to RFC 2047 encoded-word format, leaving them undeliverable. (RFC 2047 prohibits use of encoded-word in an addr-spec.) This change raises a ValueError when attempting to flatten an EmailMessage with a non-ASCII addr-spec and a policy with `utf8=False`. (Exception: If the non-ASCII address originated from parsing a message, it will be flattened as originally parsed, without error.)

Non-ASCII email addresses are supported when using a policy with `utf8=True` (such as email.policy.SMTPUTF8) under RFCs 6531 and 6532.

Non-ASCII email address domains (but not localparts) can also be used with non-SMTPUTF8 policies by encoding the domain as an IDNA A-label. (The email package does not perform this encoding, because it cannot know whether the caller wants IDNA 2003, IDNA 2008, or some other variant such as UTS-46.)


<!-- gh-issue-number: gh-83938 -->
* Issue: gh-83938
<!-- /gh-issue-number -->
<!-- gh-issue-number: gh-122476 -->
* Issue: gh-122476
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--122540.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

medmunds committed:
Stop incorrectly RFC 2047 encoding non-ASCII email addresses

Email generators had been incorrectly flattening non-ASCII email
addresses to RFC 2047 encoded-word format, leaving them undeliverable.
(RFC 2047 prohibits use of encoded-word in an addr-spec.)
This change raises a ValueError when attempting to flatten an
EmailMessage with a non-ASCII addr-spec and a policy with utf8=False.
(Exception: If the non-ASCII address originated from parsing a message,
it will be flattened as originally parsed, without error.)

Non-ASCII email addresses are supported when using a policy with
utf8=True (such as email.policy.SMTPUTF8) under RFCs 6531 and 6532.

Non-ASCII email address domains (but not localparts) can also be used
with non-SMTPUTF8 policies by encoding the domain as an IDNA A-label.
(The email package does not perform this encoding, because it cannot
know whether the caller wants IDNA 2003, IDNA 2008, or some other
variant such as UTS #46.)

medmunds committed:
Blurbs

medmunds committed:
Merge branch 'main' into fix-issues-83938-122476

medmunds committed:
Merge branch 'main' into fix-issues-83938-122476

bitdancer commented on Lib/test/test_email/test_generator.py:
```diff
@@ -288,6 +289,28 @@ def test_keep_long_encoded_newlines(self):
+                msg['To'] = address
+                expected_error = re.escape(
+                    "Non-ASCII address requires policy with utf8=True:"
+                    " '{}'".format(msg['To'].addresses[0].addr_spec)
```
I prefer to not require an exact wording for error messages.  I would write this:
```
expected_error = (
    "(?i)(?=.*non-ascii)(?=.*utf8.*True)"
    f"(?=.*{re.escape(msg['To'].addresses[0].addr_spec)})"
    )
```
to require only what I consider to be the essential parts of the message.

bitdancer commented on Lib/email/_header_value_parser.py:
```diff
@@ -2829,6 +2829,17 @@ def _refold_parse_tree(parse_tree, *, policy):
+                # Non-ASCII addr-spec came from parsed message; leave unchanged.
+                want_encoding = False
+            else:
+                raise ValueError(
```
ValueError is correct, but we don't want smtplib caching *all* ValueErrors, so we're going to need a ValueError subclass here.  How about InvalidMailboxError?

This should probably be added to the generator and/or the policy docs.

medmunds replied:
I wasn't sure whether InvalidMailboxError should subclass MessageError or HeaderWriteError, but went with the more general MessageError.

bitdancer commented on Misc/NEWS.d/next/Library/2024-07-31-17-22-10.gh-issue-83938.TtUa-c.rst:
```diff
@@ -0,0 +1,3 @@
+Stop incorrectly using RFC 2047 "encoded words" for email addresses with
+non-ASCII characters when email.generator is called using a policy with
+``utf8=False``.
```
This needs to mention the new error that will be raised.  Here is my suggestion:

The :mod:`email` module no longer incorrectly encodes non-ascii characters in email addresses using :rfc:2047 encoding.  Under a policy with `utf8=True` this means the addresses will be correctly passed through.  Under a policy with `utf8=False`, attempting to serialize a message with non-ASCII email addresses will now result in a :exc:`InvalidMailboxError`.

bitdancer reviewed (changes requested):
Your analysis is solid and the fix looks great.  We'll need a follow on PR to have smtplib handle the new error, but that should be a trivial PR.

medmunds committed:
Merge branch 'main' into fix-issues-83938-122476

medmunds committed:
fixup! Stop incorrectly RFC 2047 encoding non-ASCII email addresses

- Incorporate PR review feedback
- Improve docs

bitdancer commented on Lib/email/_header_value_parser.py:
```diff
@@ -2837,6 +2837,17 @@ def _refold_parse_tree(parse_tree, *, policy):
+            else:
+                raise errors.InvalidMailboxError(
+                    "Non-ASCII address requires policy with utf8=True:"
+                    " '{}'".format(part)
```
Would f"Non-ASCII mailbox {str(part)!r} is invalid under current policy setting (utf8=False)" be clearer do you think?

bitdancer commented on Misc/NEWS.d/next/Library/2024-07-31-17-23-06.gh-issue-122476.TtUa-c.rst:
```diff
@@ -0,0 +1,5 @@
+The :mod:`email` module no longer incorrectly encodes non-ASCII characters
```
Duplicate news item.

medmunds replied:
I've updated the blurbs to address the two issues and offer specific advice for each.

If there's some incantation that allows a single news item to cover both, happy to change to that. (But I'm guessing that requires someone who's authorized to argue with the Bedevere bot, which I am not.)

bitdancer replied:
Fair enough.  I don't know how to argue with the bot either, and as you point out the bugs are subtly different ;)

bitdancer reviewed (changes requested):
Sorry it took me so long to get back to this.

medmunds committed:
Merge branch 'main' into fix-issues-83938-122476

medmunds committed:
fixup! Stop incorrectly RFC 2047 encoding non-ASCII email addresses

- Incorporate PR feedback
- Tailor blurbs to individual issues

bitdancer committed:
Bump versionadded to 3.15

bitdancer committed:
Merge branch 'main' into fix-issues-83938-122476

bitdancer committed:
fix inadvertent typo

bitdancer reviewed (changes requested):
Well, it turns out that this bug goes a bit deeper than apparent. `addr_spec` was already supposed to be prevented from being encoded, but the logic around as_ew_allowed is broken: it doesn't "look inside" the token it is encoded to make sure *all* of it can be encoded.  The fix in this PR also has its problems, as proven by the test failures after merge with main: want_encoding can be set when what is wanted is wrapping, rather than *needing* encoding.  Those should probably have been two separate flags...but at this point the folder is a bit of a patchwork mess.  (I plan to fix that...someday).

If we fix as_eq_allowed handling I think it will make those test failures go away, as well as simplifying the logic for raising the error, but it won't be certain until we try it.

There is another test failure as well, but I think that's not a bug, but rather a change in  behavior that the test needs to adapt to.  I haven't proven that yet, though :)

bitdancer committed:
Remove incomplete fix and temporarily disable new tests.

bitdancer committed:
Move mime-parameter folding to top of loop.

The mime parameter folder doesn't make use of the encoding check
done be the code that is now below it, it does its own.  So it
makes more sense to take that branch first.  This will simplify
subsequent changes.

bitdancer committed:
Fix the bug, update the doc changes.

This is a more complete fix, covering any syntax part where encoded
words are not permitted, and the doc changes are adjusted accordingly.
There is also no need for a new exception, since HeaderWriteError
already exists.

The fix itself is to use a separate code loop to fold parts that
may not have encoded words, guaranteeing that we do not do incorrect
encoding.  This opens a door to simplifying the main folding loop,
but that is a much bigger refactoring job better left for another time.

bitdancer committed:
Add some tests where the local part is folded.

Behavior when folding in parts versus rendering on one line takes
different code paths, so make sure both work.

medmunds committed:
Update doc changes

Fix typo; bump versionchanged.

medmunds committed:
Fix reference in docs

medmunds reviewed (commented):
@bitdancer this LGTM. I pushed a couple of fixes to the docs.

I'd be inclined to keep the versionchanged note. (Unless there's some style guide policy against that for bug fixes.) There seems to be a popular misconception—in many email libraries and systems—that any MIME header with non-ASCII chars can just be run through rfc2047. The versionchanged note makes it clear dropping rfc2047 here was a deliberate fix in Python's email package, and might help discourage future requests to add it back in.

bitdancer committed:
Merge branch 'main' into fix-issues-83938-122476

medmunds committed:
Fix typo in docs

medmunds committed:
Add whatsnew item
